### PR TITLE
Enable C# 8 Nullable Reference Types

### PR DESCRIPTION
**Pull request by RehanSaeed**
Fixes https://github.com/RehanSaeed/Serilog.Exceptions/issues/175

### CODE REVIEW

File: Source/Serilog.Exceptions/Destructurers/ReflectionTypeLoadExceptionDestructurer.cs [hunk 1]
```diff
@@ -28,9 +27,29 @@ public class ReflectionTypeLoadExceptionDestructurer : ExceptionDestructurer
             {
                 propertiesBag.AddProperty(
                     nameof(ReflectionTypeLoadException.LoaderExceptions),
-                    reflectionTypeLoadException.LoaderExceptions.Select(destructureException).ToList());
+                    GetLoaderExceptionsValue(reflectionTypeLoadException.LoaderExceptions, destructureException));
```
Comment (RehanSaeed) [inline Source/Serilog.Exceptions/Destructurers/ReflectionTypeLoadExceptionDestructurer.cs:30]: Code changed here to stop using Linq and fix potential null reference exceptions.

File: Source/Serilog.Exceptions/Destructurers/TaskCanceledExceptionDestructurer.cs [hunk 2]
```diff
@@ -22,15 +22,18 @@ public class TaskCanceledExceptionDestructurer : OperationCanceledExceptionDestr
...
+            Func<Exception, IReadOnlyDictionary<string, object?>?> destructureException)
         {
 #pragma warning disable CA1062 // Validate arguments of public methods
             base.Destructure(exception, propertiesBag, destructureException);
 
             var taskCancelledException = (TaskCanceledException)exception;
-            propertiesBag.AddProperty(
-                nameof(TaskCanceledException.Task),
-                DestructureTask(taskCancelledException.Task, destructureException));
+            if (taskCancelledException.Task is not null)
```
Comment (RehanSaeed) [inline Source/Serilog.Exceptions/Destructurers/TaskCanceledExceptionDestructurer.cs:31]: Check for potential null reference exception.

File: Source/Serilog.Exceptions/DictionaryExtensions.cs [hunk 3]
```diff
@@ -13,15 +13,22 @@ internal static class DictionaryExtensions
...
-        public static Dictionary<string, object> ToStringObjectDictionary(this IDictionary dictionary)
+        public static Dictionary<string, object?> ToStringObjectDictionary(this IDictionary dictionary)
         {
-            var result = new Dictionary<string, object>(dictionary.Count);
+            var result = new Dictionary<string, object?>(dictionary.Count);
 
             foreach (var key in dictionary.Keys)
             {
-                var value = dictionary[key];
+                if (key is not null)
```
Comment (RehanSaeed) [inline Source/Serilog.Exceptions/DictionaryExtensions.cs:22]: Check for potential null reference exception.

File: Tests/Serilog.Exceptions.Test/Destructurers/ReflectionBasedDestructurerTest.cs [hunk 4]
```diff
@@ -428,66 +432,66 @@ private static Exception ThrowAndCatchException(Action throwingAction)
...
+            return null!; // We should never reach this line.
         }
 
         private static ReflectionBasedDestructurer CreateReflectionBasedDestructurer() =>
             new(10);
 
         public class MyObject
         {
-            public string Foo { get; set; }
+            public string? Foo { get; set; }
```
Comment (krajek) [inline Tests/Serilog.Exceptions.Test/Destructurers/ReflectionBasedDestructurerTest.cs:443]: Maybe we could add tests for non-nullable references as well?
Just for the sake of completeness.

File: Tools/ExceptionFinderTool/Program.cs [hunk 5]
```diff
@@ -17,7 +17,9 @@ public static void Main()
         {
             var stringBuilder = new StringBuilder();
 
-            foreach (var exceptionType in GetAllAssemblies()
+            foreach (var exceptionType in AppDomain
+                .CurrentDomain
+                .GetAssemblies()
```
Comment (krajek) [inline Tools/ExceptionFinderTool/Program.cs:22]: Is this change relevant to non-nullability?

File: Tests/Serilog.Exceptions.Test/Filters/CompositeExceptionPropertyFilterTest.cs [hunk 6]
```diff
@@ -19,65 +19,54 @@ public class CompositeExceptionPropertyFilterTest
...
 
-            // Assert
             Assert.Contains("index 0", ex.Message, StringComparison.Ordinal);
         }
 
         [Fact]
         public void ShouldFilterTheProperty_WhenFirstFilterFilters_Filters()
         {
-            // Arrange
             var filterA = new IgnorePropertyByNameExceptionFilter("A");
```
Comment (krajek) [inline Tests/Serilog.Exceptions.Test/Filters/CompositeExceptionPropertyFilterTest.cs:31]: Nitpicking: If you want to remove AAA annotations, let's do it consistently everywhere in other PR.

**Review by krajek — COMMENTED** [review]
Looks good, almost everything is nullable apart from incoming exception to be destructured, properties bag, and a string key in a dictionary. Nice improvement

File: Tests/Serilog.Exceptions.Test/Filters/CompositeExceptionPropertyFilterTest.cs (hunk 6 above)
Comment (RehanSaeed) [reply Tests/Serilog.Exceptions.Test/Filters/CompositeExceptionPropertyFilterTest.cs:31]: Yeah, fair enough. Will add back in.

File: Tests/Serilog.Exceptions.Test/Destructurers/ReflectionBasedDestructurerTest.cs (hunk 4 above)
Comment (RehanSaeed) [reply Tests/Serilog.Exceptions.Test/Destructurers/ReflectionBasedDestructurerTest.cs:443]: What do you have in mind?

File: Tools/ExceptionFinderTool/Program.cs (hunk 5 above)
Comment (RehanSaeed) [reply Tools/ExceptionFinderTool/Program.cs:22]: Could have done this in another PR. Removed some unused code only relevant to `NETCOREAPP1_1`.

File: Source/Serilog.Exceptions/Destructurers/ReflectionBasedDestructurer.cs [hunk 7]
```diff
@@ -95,12 +95,13 @@ public ReflectionBasedDestructurer(int destructuringDepth)
...
-        private static string GetOrGenerateRefId(ref int nextCyclicRefId, IDictionary<string, object> destructuredObject)
+        private static string? GetOrGenerateRefId(ref int nextCyclicRefId, IDictionary<string, object?> destructuredObject)
         {
-            string refId;
+            string? refId;
             if (destructuredObject.ContainsKey(IdLabel))
             {
-                refId = (string)destructuredObject[IdLabel];
+                var objectValue = destructuredObject[IdLabel];
+                refId = (string?)objectValue;
```
[the remark this answers is not in the document]
Comment (krajek) [reply Source/Serilog.Exceptions/Destructurers/ReflectionBasedDestructurer.cs]: I guess it is just necessary because `refId` is a `string` and `destructuredObject` dictionary contains `object` values.
For this particular special key `IdLabel` we know for sure that it contains a string because we put it there ourselves.

[the remark this answers is not in the document]
Comment (RehanSaeed) [reply Source/Serilog.Exceptions/Destructurers/ReflectionBasedDestructurer.cs]: Is there a better way we can handle that?

File: Tests/Serilog.Exceptions.Test/Filters/CompositeExceptionPropertyFilterTest.cs (hunk 6 above)
Comment (RehanSaeed) [reply Tests/Serilog.Exceptions.Test/Filters/CompositeExceptionPropertyFilterTest.cs]: Fixed

**Comment by RehanSaeed** [comment]
@krajek Are we good to go?

File: Source/Serilog.Exceptions/Destructurers/ReflectionBasedDestructurer.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (krajek) [reply Source/Serilog.Exceptions/Destructurers/ReflectionBasedDestructurer.cs]: I guess a natural way to improve (at least in one dimension) would be to introduce a layer of abstraction and replace `object` values with some explicit domain representation of `DestructuredValue` and `ReferenceId`. 

Other than that, my call is that such "low level" code is well placed in an internal class (instead of public API). Of course, I am open to suggestions.

File: Tests/Serilog.Exceptions.Test/Destructurers/ReflectionBasedDestructurerTest.cs (hunk 4 above)
Comment (krajek) [reply Tests/Serilog.Exceptions.Test/Destructurers/ReflectionBasedDestructurerTest.cs:443]: It seems that you have changed all properties in all test classes to nullable references.
I thought it would be good to have `MyObject` and `MyObjectWithNonNullable` which would have the same properties just non-nullable.
The purpose is to make sure the API works now and in the future without any issues for both cases.

**Comment by krajek** [comment]
I have no more issues with the code, thanks for the work you put into it.

File: Source/Serilog.Exceptions/Destructurers/ReflectionBasedDestructurer.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (RehanSaeed) [reply Source/Serilog.Exceptions/Destructurers/ReflectionBasedDestructurer.cs]: I think that would make a good subsequent PR,

File: Tests/Serilog.Exceptions.Test/Destructurers/ReflectionBasedDestructurerTest.cs (hunk 4 above)
Comment (RehanSaeed) [reply Tests/Serilog.Exceptions.Test/Destructurers/ReflectionBasedDestructurerTest.cs:443]: My understanding is that nullable `?` and `!` has no effect on the running of the code or on reflection.
